src/librustc_trait_selection/traits/wf.rs

```diff
@@ -522,9 +522,15 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
                     // anyway, except via auto trait matching (which
                     // only inspects the upvar types).
                     walker.skip_current_subtree(); // subtree handled below
-                    for upvar_ty in substs.as_closure().upvar_tys() {
-                        // FIXME(eddyb) add the type to `walker` instead of recursing.
-                        self.compute(upvar_ty.into());
+                    let ty = self.infcx.shallow_resolve(substs.as_closure().upvar_tuple_ty());
```

Comment:
this can just be `self.compute(substs.as_closure().upvar_tuple_ty())`, I believe, as inference variables will already be resolved in that recursive call